subctl/diagnore-pre-install.md

```diff
@@ -0,0 +1,113 @@
+
+* Broker info file.
+* Cable driver ports.
+* Is load balancer being deployed.
+
+Some of these checks can always be done as part of `join`, while others can be done only when using a flag, e.g. `--check-infra`.
+
+## Design Details
+
+We can reuse existing code in `subctl diagnose` commands as is, or refactor as necessary:
+
+* Gateway node (New check)
+  * Check that we have at least one node labeled as gateway.
+  * This check is fundamental for firewall checking.
+    * Unless we're changing deployment model not to use labeled gateway, in which case those checks will have to be updated anyhow.
+  * Kube Proxy Mode (Reuse as is)
+    * Run "ip a s kube-ipvs0" on a gateway node and check that the interface isn't there.
+    * "The cluster is deployed with kube-proxy ipvs mode which Submariner does not support"
+* Broker connectivity (New check)
+  * Make sure the broker is reachable from the cluster.
+* K8s Version (Reuse as is)
+  * Query K8s and check we have a minimal supported version (1.17).
+* CNI (Reuse if possible with refactor)
+  * Check if CNI as detected by us is supported (from a list of plugins).
+    * Reuse submariner-operator/pkg/discovery/network -> Discover
+  * Warn that Submariner might not work if the CNI not detected (but don't fail the check).
+  * Extra checks for:
+    * OVN-K:
+      * Version of NB DB is supported.
+    * Calico:
+      * IP Pools used on node labeled as gateway are disabled.
+* Overlapping CIDRs (New check)
+  * Check that the CIDRs aren't overlapping with any known CIDRs from the broker.
+  * If the clusterset is using globalnet, skip this check.
+* Firewall ports (Might be possible to reuse, at least part of it)
+  * Internal VxLAN port:
+    * Skipped on OVN-K
+    * Would be possible to do with:
+      * Spawn on a gateway labeled node a pod running `tcpdump -ln -c 5 -i any udp and src port 9898 and dst port 4500`
+      * Spawn on non-gateway a pod running `for i in $(seq 10); do timeout 2 nc -u -p 9898 <gatewayPodIP> 4500; done`
+  * Public tunnel port:
+    * Skipped when deploying with Load Balancer
+    * Quite problematic to reuse as we don't necessarily have the public IP, we would need to figure out how to deduce it.
+    * Also the existing diagnose check requires a second cluster...
+      * Although we could try to initiate from the same cluster to an external IP, but the success depends on the routing policy.
+    * Spawn on a gateway labeled node on the local cluster a pod running
+      `tcpdump -ln -Q in -A -s 100 -i any udp and dst port <dest port> | grep '<client message>'`
+      where the client message is a random string (part of UUID).
+    * Spawn on a non gateway node on the local cluster running
+      `echo <client message> | for i in $(seq 5); do timeout 2 nc -n -p 9898 -u <gateway ip> <dest port>; done`
+      where the gateway ip is the public IP of the pod running the "sniffer"
```

Comment:
Unfortunately, assuming that gateway-ip is the public-ip does not work all the time. In particular, it will not work for onPrem clusters which are behind the same Router and will also not work for disconnected/air-gapped clusters.

Reply:
We can handle air-gapped clusters when we have proper support for that.

As for on prem clusters behind NAT, what do you propose?
Won't they fail the `subctl diagnose firewall` check as well in this case?